### PR TITLE
Camera following MainDude when out of a bounding box

### PR DESCRIPTION
**Pull request by Bedzior**
What I'd like to see in this PR:
* [x] adjust the bounding box to behave like in the original game
* [x] controls' remap, to take into account the fact camera is now bound to a moving object

Stopping the camera at level's border might require small refactoring.

EDIT: I'll drop the topics for now, let's not do too much.

### CODE REVIEW

**Comment by dbeef** [comment]
Will review once the bounding box is added.

**Comment by Bedzior** [comment]
camera

File: src/camera/src/Camera.cpp [hunk 1]
```diff
@@ -8,6 +8,13 @@
 
 Camera *Camera::_camera = nullptr;
 
+Camera::Camera()
+    : _bounding_x(3.f)
```
Comment (dbeef) [inline src/camera/src/Camera.cpp:12]: Tune the values so they would correspond to the original ones. Right now the bounding box is too big.

File: src/camera/interface/Camera.hpp [hunk 2]
```diff
@@ -1,14 +1,16 @@
 #pragma once
 
 #include <cstdint>
+#include <cmath>
```
Comment (dbeef) [inline src/camera/interface/Camera.hpp:4]: Is it needed here?

File: src/camera/src/Camera.cpp (hunk 1 above)
Comment (Bedzior) [reply src/camera/src/Camera.cpp]: It's tuned to 0.5% of the original values, and for the time being it's difficult to be pixel-perfect.

**Comment by Bedzior** [comment]
One issue is, I cannot run and jump facing right 🤔

File: src/game-loop/src/main-dude/states/MainDudeBaseState.cpp [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+{
+    if (input.circle())
+    {
+        main_dude._physics.add_velocity(0.0f, -MainDude::JUMP_SPEED);
+        return &main_dude._states.jumping;
+    }
+    return nullptr;
+}
+
+MainDudeBaseState* MainDudeBaseState::handle_exiting(MainDude& main_dude, const Input& input)
```
Comment (dbeef) [inline src/game-loop/src/main-dude/states/MainDudeBaseState.cpp:27]: Why base state should be aware of something that concerns `MainDudeExitingState`?

File: src/game-loop/src/main-dude/states/MainDudeBaseState.cpp [hunk 4]
```diff
@@ -0,0 +1,51 @@
+#include "main-dude/states/MainDudeBaseState.hpp"
+#include "main-dude/MainDude.hpp"
+#include "Input.hpp"
+
+void MainDudeBaseState::handle_movement(MainDude& main_dude, const Input& input, float speed)
+{
+    if (input.left())
```
Comment (dbeef) [inline src/game-loop/src/main-dude/states/MainDudeBaseState.cpp:7]: The whole idea of having main dude split into states is to contain logics in there.

Comment (dbeef) [reply src/game-loop/src/main-dude/states/MainDudeBaseState.cpp:7]: Including handling input.

Comment (Bedzior) [reply src/game-loop/src/main-dude/states/MainDudeBaseState.cpp:7]: All right, I will drop the subject of input refactoring _for now_

File: src/game-loop/src/main-dude/states/MainDudeBaseState.cpp (hunk 3 above)
Comment (Bedzior) [reply src/game-loop/src/main-dude/states/MainDudeBaseState.cpp]: This abomination never happend

**Comment by dbeef** [comment]
2020-05-23-120221_1920x1080_scrot

Tearing artifacts when ran on an emulator and natively on Linux.

**Comment by Bedzior** [comment]
> 
> 
> 2020-05-23-120221_1920x1080_scrot
> 
> Tearing artifacts when ran on an emulator and natively on Linux.

Had this once but couldn't reproduce, luckily we worked this out 👍

**Pull request merged**